cns/restserver/ipam.go

```diff
@@ -434,26 +563,53 @@ func (service *HTTPRestService) unassignIPConfig(ipconfig cns.IPConfigurationSta
-			logger.Printf("[releaseIPConfig] Releasing IP %+v for pod %+v", ipconfig.IPAddress, podInfo)
-			_, err := service.unassignIPConfig(ipconfig, podInfo)
-			if err != nil {
-				return fmt.Errorf("[releaseIPConfig] failed to mark IPConfig [%+v] as Available. err: %v", ipconfig, err)
+	ipsToBeReleased := make([]cns.IPConfigurationStatus, len(service.PodIPIDByPodInterfaceKey[podInfo.Key()]))
+	numIPsToRelease := len(service.PodIPIDByPodInterfaceKey[podInfo.Key()])
+
+	for i, ipID := range service.PodIPIDByPodInterfaceKey[podInfo.Key()] {
+		if ipID != "" {
+			if ipconfig, isExist := service.PodIPConfigState[ipID]; isExist {
+				ipsToBeReleased[i] = ipconfig
+			} else {
+				logger.Errorf("[releaseIPConfig] Failed to get release ipconfig %+v and pod info is %+v. Pod to IPID exists, but IPID to IPConfig doesn't exist, CNS State potentially corrupt",
+					ipconfig.IPAddress, podInfo)
+				//nolint:goerr113 // return error
+				return fmt.Errorf("[releaseIPConfig] releaseIPConfig failed. IPconfig %+v and pod info is %+v. Pod to IPID exists, but IPID to IPConfig doesn't exist, CNS State potentially corrupt",
+					ipconfig.IPAddress, podInfo)
 			}
-			logger.Printf("[releaseIPConfig] Released IP %+v for pod %+v", ipconfig.IPAddress, podInfo)
 		} else {
-			logger.Errorf("[releaseIPConfig] Failed to get release ipconfig %+v and pod info is %+v. Pod to IPID exists, but IPID to IPConfig doesn't exist, CNS State potentially corrupt",
-				ipconfig.IPAddress, podInfo)
-			return fmt.Errorf("[releaseIPConfig] releaseIPConfig failed. IPconfig %+v and pod info is %+v. Pod to IPID exists, but IPID to IPConfig doesn't exist, CNS State potentially corrupt",
-				ipconfig.IPAddress, podInfo)
+			logger.Errorf("[releaseIPConfig] SetIPConfigAsAvailable ignoring request to release, no allocation found for pod [%+v]", podInfo)
+			break
 		}
-	} else {
-		logger.Errorf("[releaseIPConfig] SetIPConfigAsAvailable ignoring request to release, no allocation found for pod [%+v]", podInfo)
-		return nil
 	}
+
+	for i := range ipsToBeReleased {
+		logger.Printf("[releaseIPConfig] Releasing IP %+v for pod %+v", ipsToBeReleased[i].IPAddress, podInfo)
+		_, err := service.unassignIPConfig(ipsToBeReleased[i], podInfo)
+		if err != nil {
+			return fmt.Errorf("[releaseIPConfig] failed to mark IPConfig [%+v] as Available. err: %w", ipsToBeReleased[i], err)
+		}
+		logger.Printf("[releaseIPConfig] Released IP %+v for pod %+v", ipsToBeReleased[i].IPAddress, podInfo)
+		if i == len(service.state.ContainerStatus)-1 {
+			return nil
+		}
+	}
+
+	// if we were able to get at least one IP but not all of the desired IPs
+	logger.Printf("[releaseIPConfig] Released IP %+v for pod %+v", len(ipsToBeReleased), podInfo)
+	if len(ipsToBeReleased) < numIPsToRelease || (numIPsToRelease < len(service.state.ContainerStatus) && numIPsToRelease != 0) {
+		for i := range ipsToBeReleased {
+			if ipsToBeReleased[i].ID != "" {
+				err := service.assignIPConfig(ipsToBeReleased[i], podInfo)
```

Comment:
I believe we should not try to assign back those ips. Whtever we have released can go back to free pool

Reply:
I thought one of the things we needed was to not partially release IPs? If we have two IPs to a pod and one is released and the other isn't would we consider that to be ok if one is still assigned? 
If the v4 one is released and the v6 isn't than having an extra v4 IP in the pool wouldn't really allow you to create more pods since you would still be short a v6 IP and it would say waiting for more IPs.

Reply:
revised